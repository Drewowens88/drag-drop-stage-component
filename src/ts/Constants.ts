
export const MIN_SIZE = 20;
export const STICK_DISTANCE = 5;
