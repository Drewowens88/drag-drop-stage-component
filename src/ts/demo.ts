import {Stage} from './index';

// find the empty iframe in the page
const iframe = document.querySelector('#iframe') as HTMLIFrameElement;
const output = document.querySelector('#output') as HTMLInputElement;
// write some content in the iframe
iframe.contentDocument.write(document.querySelector('#random-content').innerHTML);
// create the Stage class
new Stage(iframe, iframe.contentDocument.querySelectorAll('.stage-element'), {
  onDrop: (selectables) => {
    const dropElement = selectables[0].dropZone.parent;
    const str = `${ selectables.length } elements have been droped to ${ dropElement.tagName.toLocaleLowerCase() }${ dropElement.id ? '#' + dropElement.id : '' }${ Array.from(dropElement.classList).map(c => '.' + c).join('') }`;
    console.log(str);
    output.value = str;
  },
  onDrag: (selectables, boundingBox) => {
    const str = `${ selectables.length } elements are being draged`;
    console.log(str);
    output.value = str;
  },
  onResize: (selectables, boundingBox) => {
    const str = `${ selectables.length } elements have been resizeed to ${JSON.stringify(boundingBox)}`;
    console.log(str);
    output.value = str;
  },
  onDraw: (selectables, boundingBox) => {
    const str = `Drawing region: ${JSON.stringify(boundingBox)}`;
    console.log(str);
    output.value = str;
  },
  onSelect: (selectables) => {
    const str = `${ selectables.length } elements have been (un)selected`;
    console.log(str);
    output.value = str;
  },
});
