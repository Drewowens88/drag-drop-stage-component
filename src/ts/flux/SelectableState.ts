import * as DomMetrics from '../utils/DomMetrics';
import { SelectableState } from '../Types';

const UPDATE = 'SELECTABLE_UPDATE';
const RESET = 'SELECTABLE_RESET';
const CREATE = 'SELECTABLE_CREATE';
const DELETE = 'SELECTABLE_DELETE';

export const updateSelectables = (selectables: Array<SelectableState>, preventDispatch: boolean = false) => ({
  type: UPDATE,
  selectables,
  preventDispatch,
});
export const resetSelectables = () => ({
  type: RESET,
});
export const createSelectable = (selectable: SelectableState) => ({
  type: CREATE,
  selectable,
});
export const deleteSelectable = (selectable: SelectableState) => ({
  type: DELETE,
  selectable,
});

export const selectables = (state=[], action) => {
  switch(action.type) {
    case CREATE:
      return [
        ...state,
        action.selectable,
      ];
    case RESET:
      return [
        ...state,
        [],
      ];
    case DELETE:
      return state.filter((selectable: SelectableState) => selectable.el !== action.selectable.el);
    case UPDATE:
      return state.map((selectable: SelectableState) => action.selectables.find(s => s.el === selectable.el) || selectable);
    default:
      return state;
  }
};
