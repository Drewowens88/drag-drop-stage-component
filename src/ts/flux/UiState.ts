import * as types from '../Types'

export const UI_SET_MODE = 'UI_SET_MODE';
export const setMode = (mode: types.UiMode) => ({
  type: UI_SET_MODE,
  mode,
});

export const UI_SET_REFRESHING = 'UI_SET_REFRESHING';
export const setRefreshing = (refreshing: boolean) => ({
  type: UI_SET_REFRESHING,
  refreshing,
});

export const UI_SET_CATCHING_EVENTS = 'UI_SET_CATCHING_EVENTS';
export const setCatchingEvents = (catchingEvents: boolean) => ({
  type: UI_SET_CATCHING_EVENTS,
  catchingEvents,
});

/**
 * reducer
 */
export const ui = (state=getDefaultState(), action) => {
  switch(action.type) {
    case UI_SET_MODE:
      return {
        ...state,
        mode: action.mode,
      }
    case UI_SET_REFRESHING:
      return {
        ...state,
        refreshing: action.refreshing,
      }
    case UI_SET_CATCHING_EVENTS:
      return {
        ...state,
        catchingEvents: action.catchingEvents,
      }
    default:
      return state;
  }
};

export const getDefaultState = () => {
  return {
    mode: types.UiMode.NONE,
    refreshing: false,
    catchingEvents: true,
  };
}
