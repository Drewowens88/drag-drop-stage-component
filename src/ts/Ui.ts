import {StageStore} from './flux/StageStore';
import { SelectableState, State, MouseState, MouseData, ScrollData, ElementMetrics } from './Types';
import { addEvent } from './utils/Events';
import * as DomMetrics from './utils/DomMetrics';

interface Box {
  selectable: SelectableState;
  ui: HTMLElement;
}

export class Ui {

  overlay: HTMLIFrameElement;
  boxes: Array<Box> = [];

  constructor(iframe: HTMLIFrameElement, store: StageStore) {
    const doc = DomMetrics.getDocument(iframe);
    const win = DomMetrics.getWindow(doc);

    // create the overlay
    doc.body.style.overflow = 'hidden';
    if(!!this.overlay && !!this.overlay.parentElement) {
      this.overlay.parentElement.removeChild(this.overlay); // could use less supported this.overlay.remove()
    }
    this.overlay = doc.createElement('iframe');
    doc.body.appendChild(this.overlay);

    // init iframe
    this.resize(DomMetrics.getBoundingBoxDocument(iframe), win.getComputedStyle(iframe).zIndex);
    this.update(store.getState().selectables, this.getScrollData(iframe));

    // overlay content
    this.overlay.contentDocument.head.innerHTML = `
      <style>
        body {
          overflow: hidden;
          margin: -5px;
        }

        body.dragging-mode .box.not-selected,
        body.resizing-mode .box.not-selected { display: none; }

        .selected.box, .box:hover {
          border: 1px solid rgba(0, 0, 0, .5);
        }
        .selected.box:before,
        .box:hover:before {
          content: " ";
          position: absolute;
          z-index: -1;
          top: 1px;
          left: 1px;
          right: 1px;
          bottom: 1px;
          border: 1px solid rgba(255, 255, 255, .3);
        }
        .not-selectable,
        .not-selected .handle { display: none; }

        .handle {
          position: absolute;
          border: 1px solid rgba(0, 0, 0, .5);
          background-color: rgba(255, 255, 255, 1);
          width: 5px;
          height: 5px;
          border-radius: 2.5px;
        }
        .handle-nw { top: -4px; left: -4px; }
        .not-resizeable-nw .handle-nw { display: none; }

        .handle-ne { top: -4px; right: -4px; }
        .not-resizeable-ne .handle-ne { display: none; }

        .handle-sw { bottom: -4px; left: -4px; }
        .not-resizeable-sw .handle-sw { display: none; }

        .handle-se { bottom: -4px; right: -4px; }
        .not-resizeable-se .handle-se { display: none; }


        .region-marker {
          background-color: rgba(0, 0, 0, .1);
          border: 1px solid rgba(255, 255, 255, .5);
          display: flex;
          position: absolute;
          left: 0;
          top: 0;
          min-width: 1px;
          min-height: 1px;
        }
    `;
    this.unsubscribeAll.push(
      addEvent(win, 'resize', () => {
        this.resize(DomMetrics.getBoundingBoxDocument(iframe), win.getComputedStyle(iframe).zIndex)
        this.update(store.getState().selectables, this.getScrollData(iframe));
      }),
      store.subscribe(
        (selectables: Array<SelectableState>) => this.update(selectables, this.getScrollData(iframe)),
        (state: State) => state.selectables,
      ),
      store.subscribe(
        (state: MouseState, prevState: MouseState) => this.onMouseChanged(state, prevState),
        (state:State) => state.mouse
      ),
    );
  }

  private unsubscribeAll: Array<() => void> = [];
  cleanup() {
    this.unsubscribeAll.forEach(u => u());
  }

  private getScrollData(iframe: HTMLIFrameElement): ScrollData {
    return {
      x: iframe.contentWindow.document.scrollingElement.scrollWidth,
      y: iframe.contentWindow.document.scrollingElement.scrollHeight,
    };
  }
  private resize(bb: ClientRect, zIndex: string) {
    this.overlay.style.border = 'none';
    this.overlay.style.zIndex = ((parseInt(zIndex) || 0) + 1).toString();
    this.overlay.style.backgroundColor = 'transparent';
    this.overlay.style.position = 'absolute';
    this.overlay.style.top = bb.top + 'px';
    this.overlay.style.left = bb.left + 'px';
    this.overlay.style.width = bb.width + 'px';
    this.overlay.style.height = bb.height + 'px';
  }
  onMouseChanged(state: MouseState, prevState: MouseState) {
    if(state.scrollData.x !== prevState.scrollData.x || state.scrollData.y !== prevState.scrollData.y) {
      DomMetrics.setScroll(this.overlay.contentDocument, state.scrollData);
    }
    if(state.cursorData.cursorType !== prevState.cursorData.cursorType) {
      this.overlay.contentDocument.body.style.cursor = state.cursorData.cursorType;
    }
  }
  update(selectables: Array<SelectableState>, scrollData: ScrollData) {
    //  update scroll
    this.overlay.contentDocument.body.style.overflow = 'auto';
    this.overlay.contentDocument.body.style.width = scrollData.x + 'px';
    this.overlay.contentDocument.body.style.height = scrollData.y + 'px';

    // remove the UIs that have no corresponding element in the stage
    this.boxes
    .filter(r => !selectables.find(s => r.selectable.el === s.el))
    .forEach(r => r.ui.parentElement.removeChild(r.ui));

    // remove the boxes
    this.boxes = this.boxes
    .filter(r => selectables.find(s => r.selectable.el === s.el))

    // add the missing boxes
    this.boxes = this.boxes.concat(
      selectables
      // only the missing ones
      .filter(s => !this.boxes.find(r => r.selectable.el === s.el))
      // create a box object
      .map(s => ({
        selectable: s,
        // append a new div to the overlay
        ui: this.overlay.contentDocument.body.appendChild(this.createBoxUi()),
      }))
    );

    // update the view
    this.boxes.map(r => this.updateBox(r, selectables.find(s => s.el === r.selectable.el)));
  }
  createBoxUi(): HTMLElement {
    const box = this.overlay.contentDocument.createElement('div');
    box.innerHTML = `
      <div class="handle handle-nw"></div>
      <div class="handle handle-ne"></div>
      <div class="handle handle-sw"></div>
      <div class="handle handle-se"></div>
    `;
    return box;
  }
  updateBox(box: Box, selectable: SelectableState): Box {
    box.selectable = selectable;
    DomMetrics.setMetrics(box.ui, {
      ...box.selectable.metrics,
      position: 'absolute',
      padding: {top: 0, left: 0, bottom: 0, right: 0},
      margin: {top: 0, left: 0, bottom: 0, right: 0},
      border: {top: 1, left: 1, bottom: 1, right: 1},
    }, false, true);
    box.ui.classList.remove(...[
      !box.selectable.selected ? 'selected' : 'not-selected',
      !box.selectable.selectable ? 'selectable' : 'not-selectable',
      !box.selectable.draggable ? 'draggable' : 'not-draggable',
      (!box.selectable.resizeable.top && !box.selectable.resizeable.left) ? 'resizeable-nw' : 'not-resizeable-nw',
      (!box.selectable.resizeable.top && !box.selectable.resizeable.right) ? 'resizeable-ne' : 'not-resizeable-ne',
      (!box.selectable.resizeable.bottom && !box.selectable.resizeable.left) ? 'resizeable-sw' : 'not-resizeable-sw',
      (!box.selectable.resizeable.bottom && !box.selectable.resizeable.right) ? 'resizeable-se' : 'not-resizeable-se',
      !box.selectable.isDropZone ? 'isDropZone' : 'not-isDropZone',
    ]);
    box.ui.classList.add(...[
      'box',
      box.selectable.selected ? 'selected' : 'not-selected',
      box.selectable.selectable ? 'selectable' : 'not-selectable',
      box.selectable.draggable ? 'draggable' : 'not-draggable',
      (box.selectable.resizeable.top && box.selectable.resizeable.left) ? 'resizeable-nw' : 'not-resizeable-nw',
      (box.selectable.resizeable.top && box.selectable.resizeable.right) ? 'resizeable-ne' : 'not-resizeable-ne',
      (box.selectable.resizeable.bottom && box.selectable.resizeable.left) ? 'resizeable-sw' : 'not-resizeable-sw',
      (box.selectable.resizeable.bottom && box.selectable.resizeable.right) ? 'resizeable-se' : 'not-resizeable-se',
      box.selectable.isDropZone ? 'isDropZone' : 'not-isDropZone',
    ]);
    return box;
  }
}