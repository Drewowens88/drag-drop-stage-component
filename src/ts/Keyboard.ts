import { StageStore } from './flux/StageStore';
import { addEvent } from './utils/Events';
import { setMode } from './flux/UiState';
import { reset } from './flux/SelectionState';
import { UiMode, Hooks } from './Types';
import { updateSelectables } from './flux/SelectableState';
import * as DomMetrics from './utils/DomMetrics';

const MOVE_DISTANCE = 5;
const SHIFT_MOVE_DISTANCE = 1;
const ALT_MOVE_DISTANCE = 10;

export class Keyboard {
  constructor(private win: Window, private store: StageStore, private hooks: Hooks) {
    // events from inside the iframe
    this.unsubscribeAll.push(
      addEvent(window, 'keydown', (e: KeyboardEvent) => this.onKeyDown(e)),
      addEvent(win, 'keydown', (e: KeyboardEvent) => this.onKeyDown(e)),
    );
  }

  private unsubscribeAll: Array<() => void> = [];
  cleanup() {
    this.unsubscribeAll.forEach(u => u());
  }

  /**
   * handle shortcuts
   */
  private onKeyDown(e: KeyboardEvent) {
    const key = e.key;
    const state = this.store.getState();
    const target = e.target as HTMLElement;

    if(state.ui.catchingEvents &&
      target.tagName.toLowerCase() !== 'input' &&
      target.tagName.toLowerCase() !== 'textarea' &&
      !target.hasAttribute('contenteditable')) {
      switch(key) {
        case 'Escape':
          if(state.ui.mode !== UiMode.NONE) {
            this.store.dispatch(setMode(UiMode.NONE));
            this.store.dispatch(reset());
          }
          break;
        case 'Enter':
          if(this.hooks.onEdit) this.hooks.onEdit();
          break;
        case 'ArrowLeft':
          this.move(-this.getDistance(e), 0);
          break;
        case 'ArrowUp':
          this.move(0, -this.getDistance(e));
          break;
        case 'ArrowRight':
          this.move(this.getDistance(e), 0);
          break;
        case 'ArrowDown':
          this.move(0, this.getDistance(e));
          break;
        default:
          return;
      }
      // only if we catched a shortcut
      e.preventDefault();
      e.stopPropagation();
    }
  }
  getDistance(e: KeyboardEvent) {
    return e.shiftKey ? SHIFT_MOVE_DISTANCE :
      e.altKey ? ALT_MOVE_DISTANCE : MOVE_DISTANCE;
  }
  move(movementX, movementY) {
    const updated = this.store.getState().selectables
    .filter(s => s.selected && this.hooks.isDraggable(s.el))
    .map(selectable => {
      if(selectable.metrics.position === 'static') {
        // move the element in the dom
        const element = selectable.el;
        if((movementX < 0 || movementY < 0) && element.previousElementSibling) {
          element.parentNode.insertBefore(element, element.previousElementSibling);
        }
        else if((movementX > 0 || movementY > 0) && element.nextElementSibling) {
          element.parentNode.insertBefore(element.nextElementSibling, element);
        }
        else {
          // nothing happened
          return null;
        }
        selectable.dropZone = {
          parent: element.parentNode as HTMLElement,
        };
        return {
          ...selectable,
          metrics: DomMetrics.getMetrics(selectable.el),
        };
      }
      return {
        ...selectable,
        metrics: {
          ...selectable.metrics,
          clientRect : {
            ...selectable.metrics.clientRect,
            top: selectable.metrics.clientRect.top + movementY,
            left: selectable.metrics.clientRect.left + movementX,
            bottom: selectable.metrics.clientRect.bottom + movementY,
            right: selectable.metrics.clientRect.right + movementX,
          },
          computedStyleRect: {
            ...selectable.metrics.computedStyleRect,
            top: selectable.metrics.computedStyleRect.top + movementY,
            left: selectable.metrics.computedStyleRect.left + movementX,
            bottom: selectable.metrics.computedStyleRect.bottom + movementY,
            right: selectable.metrics.computedStyleRect.right + movementX,
          },
        }
      };
    })
    .filter(s => !!s);
    this.store.dispatch(updateSelectables(updated));
    const domChanged = updated.filter(s => !!s.dropZone);
    if(domChanged.length > 0) {
      if(this.hooks.onDrop) this.hooks.onDrop(domChanged);
    }
  }
}
