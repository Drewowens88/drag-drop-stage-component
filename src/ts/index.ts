import * as types from './Types';
import * as Polyfill from './utils/Polyfill';
import * as DomMetrics from './utils/DomMetrics';
import {Keyboard} from './Keyboard';
import {Mouse} from './Mouse';
import * as selectionState from './flux/SelectionState';
import * as UiAction from './flux/UiState';
import {SelectablesObserver} from './observers/SelectablesObserver';
import {MouseObserver} from './observers/MouseObserver';
import {UiObserver} from './observers/UiObserver';
import {StageStore} from './flux/StageStore';
import { resetSelectables, createSelectable, updateSelectables, deleteSelectable } from './flux/SelectableState';
import { addEvent } from './utils/Events';
import { setScroll } from './flux/MouseState';
import { Ui } from './Ui';

/**
 * This class is the entry point of the library
 * @see https://github.com/lexoyo/stage
 * @class Stage
 */
export class Stage {
  protected contentWindow: Window;
  protected contentDocument: HTMLDocument;
  protected iframe: HTMLIFrameElement;
  protected hooks: types.Hooks;
  protected unsubscribeAll: Array<() => void> = [];
  protected ui: Ui;
  protected store: StageStore;

  /**
   * Init the useful classes,
   * add styles to the iframe for the UI,
   * listen to mouse events in the iframe and outside
   * @constructor
   */
  constructor(iframe: HTMLIFrameElement, elements: ArrayLike<HTMLElement>, options: types.Hooks={}) {
    // expose for client app
    window['Stage'] = Stage;

    // store the params
    this.iframe = iframe;
    this.contentWindow = this.iframe.contentWindow;
    this.contentDocument = this.iframe.contentDocument;
    this.hooks = {
      ...options, // other hooks without default values
      isSelectable: options.isSelectable || (el => el.classList.contains('selectable')),
      isDraggable: options.isDraggable || (el => el.classList.contains('draggable')),
      isDropZone: options.isDropZone || ((el) => el.classList.contains('droppable')),
      isResizeable: options.isResizeable || ((el) => el.classList.contains('resizeable')),
      useMinHeight: options.useMinHeight || ((el) => true),
      canDrop: options.canDrop || ((el, dropZone) => true),
    }

    // polyfill the iframe
    Polyfill.patchWindow(this.contentWindow);

    // create the store and populate it
    this.store = new StageStore();
    Array.from(elements).forEach(el => this.addElement(el))

    // add a UI over the iframe
    this.ui = new Ui(iframe, this.store);

    // state observers
    const selectablesObserver = new SelectablesObserver(this.contentDocument, this.ui.overlay.contentDocument, this.store, this.hooks);
    const uiObserver = new UiObserver(this.contentDocument, this.ui.overlay.contentDocument, this.store, this.hooks);
    const mouseObserver = new MouseObserver(this.contentDocument, this.ui.overlay.contentDocument, this.store, this.hooks);

    // controllers
    const mouse = new Mouse(this.contentWindow, this.ui.overlay.contentWindow, this.store, this.hooks);
    const keyboard = new Keyboard(this.ui.overlay.contentWindow, this.store, this.hooks);

    this.unsubscribeAll.push(
      () => selectablesObserver.cleanup(),
      () => uiObserver.cleanup(),
      () => mouseObserver.cleanup(),
      () => mouse.cleanup(),
      () => keyboard.cleanup(),

      // window resize
      addEvent(window, 'resize', (e: MouseEvent) => this.redraw()),
    );
  }


  /**
   * to be called before deleting the stage
   */
  cleanup() {
    this.unsubscribeAll.forEach(u => u());
    this.ui.cleanup();
    this.ui = null;
  }

  /**
   * enable/disable catching events
   */
  get catchingEvents(): boolean {
    return this.store.getState().ui.catchingEvents;
  }
  set catchingEvents(val: boolean) {
    this.store.dispatch(UiAction.setCatchingEvents(val));
  }
  /**
   * enable/disable catching events
   */
  get visible(): boolean {
    return this.store.getState().ui.mode !== types.UiMode.HIDE;
  }
  set visible(val: boolean) {
    // dispatch UI mode change
    this.store.dispatch(UiAction.setMode(val ? types.UiMode.NONE : types.UiMode.HIDE));
    // scroll may have changed
    this.setScroll({
      x: this.iframe.contentWindow.scrollX,
      y: this.iframe.contentWindow.scrollY,
    });
  }
  /**
   * force resize of UI
   */
  resizeWindow() {
    this.ui.resizeOverlay();
  }

  ///////////////////////////////////////////////////
  // Elements and metrics
  ///////////////////////////////////////////////////

  /**
   * recalculate all the metrics
   */
  redraw() {
    this.store.dispatch(UiAction.setRefreshing(true));
    this.store.dispatch(updateSelectables(this.store.getState().selectables.map(selectable => {
      return {
        ...selectable,
        metrics: DomMetrics.getMetrics(selectable.el),
      }
    })));
    this.store.dispatch(UiAction.setRefreshing(false));
  }

  /**
   * get/set the states of the selected elements
   */
  getSelection(): Array<types.SelectableState> {
    return DomMetrics.getSelection(this.store);
  }

  /**
   * get/set the states of the selected elements
   */
  setSelection(elements: Array<HTMLElement>) {
    this.store.dispatch(selectionState.set(elements.map(el => this.getState(el))));
  }

  /**
   * get/set the state for an element
   */
  getState(el: HTMLElement): types.SelectableState {
    return DomMetrics.getSelectable(this.store, el);
  }

  /**
   * get/set the state for an element
   */
  setState(el: HTMLElement, subState: {
    selected?: boolean,
    selectable?: boolean,
    draggable?: boolean,
    resizeable?: types.Direction,
    isDropZone?: boolean,
    useMinHeight?: boolean,
    metrics?: types.ElementMetrics,
  }) {
    const state = this.getState(el);
    this.store.dispatch(updateSelectables([{
      ...state,
      ...subState,
    }]));
  }

  /**
   * Add an element to the store
   */
  addElement(el: HTMLElement) {
    const boolOrObj: types.Direction | boolean = this.hooks.isResizeable(el);
    const resizeable = typeof boolOrObj === 'object' ? boolOrObj : {
      top: boolOrObj as boolean,
      left: boolOrObj as boolean,
      bottom: boolOrObj as boolean,
      right: boolOrObj as boolean,
    }
    // do not apply style change to this element
    this.store.dispatch(UiAction.setRefreshing(true));

    // create an element in the store
    this.store.dispatch(createSelectable({
      el,
      selected: false,
      selectable: this.hooks.isSelectable(el),
      draggable: this.hooks.isDraggable(el),
      resizeable,
      isDropZone: this.hooks.isDropZone(el),
      useMinHeight: this.hooks.useMinHeight(el),
      metrics: DomMetrics.getMetrics(el),
    }));
    // compute all metrics again because this element might affect others
    this.redraw();
  }

  /**
   * Remove an element from the store
   */
  removeElement(el: HTMLElement) {
    this.store.dispatch(deleteSelectable(this.store.getState().selectables.find(s => s.el === el)));
    // compute all metrics again because this element might affect others
    this.redraw();
  }


  /**
   * get the best drop zone at a given position
   * if the `el` param is provided, filter possible drop zones with the `canDrop` hook
   */
  getDropZone(x: number, y: number, el:HTMLElement = null): HTMLElement {
    const dropZones = DomMetrics.findDropZonesUnderMouse(this.contentDocument, this.store, this.hooks, x, y);
    if(!!el) return dropZones.filter(dropZone => this.hooks.canDrop(el, dropZone))[0];
    else return dropZones[0];
  }

  ///////////////////////////////////////////////////
  // Scroll
  ///////////////////////////////////////////////////

  /**
   * scroll so that the elements are visible
   */
  show(elements: Array<HTMLElement>) {
    const state = this.store.getState();
    const bb = DomMetrics.getBoundingBox(elements.map(el => state.selectables.find(s => s.el === el)));
    const initialScroll: types.ScrollData = state.mouse.scrollData;
    const scroll: types.ScrollData = DomMetrics.getScrollToShow(this.contentDocument, bb);
    if(scroll.x !== initialScroll.x || scroll.y !== initialScroll.y) {
      this.store.dispatch(setScroll(scroll));
    }
  }


  /**
   * scroll so that the elements are centered
   */
  center(elements: Array<HTMLElement>) {
    const state = this.store.getState();
    const bb = DomMetrics.getBoundingBox(
      elements
      .map(el => state.selectables.find(s => s.el === el))
      .filter(s => !!s)
    );
    const initialScroll: types.ScrollData = state.mouse.scrollData;
    const scroll: types.ScrollData = {
      x: Math.max(0, Math.round(bb.left + (bb.width/2))),
      y: Math.max(0, Math.round(bb.top + (bb.height/2))),
    };
    if(scroll.x !== initialScroll.x || scroll.y !== initialScroll.y) {
      this.store.dispatch(setScroll(scroll));
    }
  }


  /**
   * get/set the stage scroll data
   */
  setScroll(scroll: types.ScrollData) {
    const initialScroll: types.ScrollData = this.store.getState().mouse.scrollData;
    if(scroll.x !== initialScroll.x || scroll.y !== initialScroll.y) {
      this.store.dispatch(setScroll(scroll));
    }
  }


  /**
   * get/set the stage scroll data
   */
  getScroll(): types.ScrollData {
    return this.store.getState().mouse.scrollData;
  }
}
