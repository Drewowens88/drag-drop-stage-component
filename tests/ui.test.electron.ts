import * as electron from 'electron';
import {Ui} from '../src/ts/Ui';
import { StageStoreMock } from './flux/StageStoreMock';

describe('Ui', () => {

  var ui: Ui;
  var iframeTest: HTMLIFrameElement;
  var stageStoreMock: StageStoreMock;
  var iframeUi: HTMLIFrameElement;

  function resetUi() {
    iframeTest = document.querySelector('#iframe-test-id');
    ui = new Ui(iframeTest, stageStoreMock);
    jest.spyOn(ui, 'update');

    const iframes = Array.from(document.querySelectorAll('iframe'));
    iframeUi = iframes.find(i => i.id != 'iframe-test-id');
  }

  function addSelectable() {
    stageStoreMock.state = {
      ...stageStoreMock.state,
      selectables: [
        ...stageStoreMock.state.selectables,
        {
          el: null,
          selectable: true,
          selected: false,
          draggable: true,
          resizeable: true,
          isDropZone: true,
          useMinHeight: true,
          metrics: {
            position: 'static',
            proportions: 1,
            margin: {top: 0, left: 0, bottom: 0, right: 0 },
            padding: {top: 0, left: 0, bottom: 0, right: 0 },
            border: {top: 0, left: 0, bottom: 0, right: 0 },
            computedStyleRect: {top: 10000, left: 10000, bottom: 20000, right: 20000, width: 10000, height: 10000 },
            clientRect: {top: 10000, left: 10000, bottom: 20000, right: 20000, width: 10000, height: 10000 },
          },
        },
      ]
    };
    stageStoreMock.dispatch(null)
  }

  beforeEach((done) => {
    // DOM
    document.body.innerHTML = `
      <style>
        #iframe-test-id {
          position: relative;
          width: 1000px;
          height: 1000px;
          margin: 100px;
        }
      </style>
      <iframe id="iframe-test-id" />
    `;

    // Store mock
    stageStoreMock = new StageStoreMock()
    jest.spyOn(stageStoreMock, 'subscribe');
    jest.spyOn(stageStoreMock, 'dispatch');
    jest.spyOn(stageStoreMock, 'getState');

    resetUi();

    // const timeout = 10000;
    // jest.setTimeout(timeout);
    // electron.remote.getCurrentWindow().show();
    // setTimeout(() => { done() }, timeout);
    // return;
    setTimeout(() => { done() }, 0);
  });


  it('should create an iframe on top of the iframe', () => {
    expect(ui).not.toBeNull();

    const iframes = Array.from(document.querySelectorAll('iframe'));
    expect(iframes.length).toBe(2);

    expect(iframeUi).not.toBeNull();
    expect(iframeUi.getBoundingClientRect()).toMatchObject(iframeTest.getBoundingClientRect());
    expect(parseInt(window.getComputedStyle(iframeUi).zIndex)).toBe(1);
    expect(window.getComputedStyle(iframeUi).backgroundColor).toBe('rgba(0, 0, 0, 0)');

    window.dispatchEvent(new Event('resize'));
    expect(ui.update).toBeCalledTimes(1);
  });

  it('should listen to the store and update', () => {
    expect(stageStoreMock.subscribe).toBeCalledTimes(2);
    addSelectable();
    expect(ui.update).toBeCalledTimes(1);
  });

  it('should draw the UI', () => {
    var elements = Array.from(iframeUi.contentDocument.body.querySelectorAll('.box'));
    expect(elements.length).toBe(2);
    elements.forEach((el, idx) => expect(el).toMatchSnapshot('snapshot-draw-ui' + idx));

    addSelectable();

    var elements = Array.from(iframeUi.contentDocument.body.querySelectorAll('.box'));
    expect(elements.length).toBe(3);
    elements.forEach((el, idx) => expect(el).toMatchSnapshot('snapshot-draw-ui' + idx));
  });

  it('should change the mouse cursor', () => {
    // css class
    stageStoreMock.state = {
      ...stageStoreMock.state,
      mouse: {
        ...stageStoreMock.state.mouse,
        cursorData: {
          y: '',
          x: '',
          cursorType: 'alias',
        }
      }
    };
    stageStoreMock.dispatch(null, null, 1);
    expect(iframeUi.contentDocument.body.style.cursor).toBe('alias');
  });
});