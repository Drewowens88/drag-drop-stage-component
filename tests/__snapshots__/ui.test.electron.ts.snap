// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ui should draw the UI: snapshot-draw-ui0 1`] = `
<div
  class="box not-selected selectable draggable resizeable-nw resizeable-ne resizeable-sw resizeable-se isDropZone"
  style="position: absolute; top: 100px; left: 100px; width: 98px; height: 98px; border-width: 1px;"
>
  
      
  <div
    class="handle handle-nw"
  />
  
      
  <div
    class="handle handle-ne"
  />
  
      
  <div
    class="handle handle-sw"
  />
  
      
  <div
    class="handle handle-se"
  />
  
    
</div>
`;

exports[`Ui should draw the UI: snapshot-draw-ui0 2`] = `
<div
  class="box not-selected selectable draggable resizeable-nw resizeable-ne resizeable-sw resizeable-se isDropZone"
  style="position: absolute; top: 100px; left: 100px; width: 98px; height: 98px; border-width: 1px;"
>
  
      
  <div
    class="handle handle-nw"
  />
  
      
  <div
    class="handle handle-ne"
  />
  
      
  <div
    class="handle handle-sw"
  />
  
      
  <div
    class="handle handle-se"
  />
  
    
</div>
`;

exports[`Ui should draw the UI: snapshot-draw-ui1 1`] = `
<div
  class="box not-selected selectable draggable resizeable-nw resizeable-ne resizeable-sw resizeable-se isDropZone"
  style="position: absolute; top: 100px; left: 100px; width: 98px; height: 98px; border-width: 1px;"
>
  
      
  <div
    class="handle handle-nw"
  />
  
      
  <div
    class="handle handle-ne"
  />
  
      
  <div
    class="handle handle-sw"
  />
  
      
  <div
    class="handle handle-se"
  />
  
    
</div>
`;

exports[`Ui should draw the UI: snapshot-draw-ui1 2`] = `
<div
  class="box not-selected selectable draggable resizeable-nw resizeable-ne resizeable-sw resizeable-se isDropZone"
  style="position: absolute; top: 100px; left: 100px; width: 98px; height: 98px; border-width: 1px;"
>
  
      
  <div
    class="handle handle-nw"
  />
  
      
  <div
    class="handle handle-ne"
  />
  
      
  <div
    class="handle handle-sw"
  />
  
      
  <div
    class="handle handle-se"
  />
  
    
</div>
`;

exports[`Ui should draw the UI: snapshot-draw-ui2 1`] = `
<div
  class="box not-selected selectable draggable resizeable-nw resizeable-ne resizeable-sw resizeable-se isDropZone"
  style="position: absolute; top: 10000px; left: 10000px; width: 9998px; height: 9998px; border-width: 1px;"
>
  
      
  <div
    class="handle handle-nw"
  />
  
      
  <div
    class="handle handle-ne"
  />
  
      
  <div
    class="handle handle-sw"
  />
  
      
  <div
    class="handle handle-se"
  />
  
    
</div>
`;
